diff --git a/atdgen-runtime/src/oj_run.ml b/atdgen-runtime/src/oj_run.ml
index f5c1217..0da4c03 100644
--- a/atdgen-runtime/src/oj_run.ml
+++ b/atdgen-runtime/src/oj_run.ml
@@ -236,33 +236,4 @@ let write_with_adapter restore writer ob x =
   let ast' = restore ast in
   Yojson.Safe.to_outbuf ob ast'
 
-(* We want an identity function that is not inlined *)
-type identity_t = { mutable _identity : 'a. 'a -> 'a }
-let identity_ref = { _identity = (fun x -> x) }
-let identity x = identity_ref._identity x
-
-(*
-  Checking at runtime that our assumptions on unspecified compiler behavior
-  still hold.
-*)
-
-type t = {
-  _a : int option;
-  _b : int;
-}
-
-let create () =
-  { { _a = None; _b = Array.length Sys.argv } with _a = None }
-
-let test () =
-  let r = create () in
-  let v = Some 17 in
-  Obj.set_field (Obj.repr r) 0 (Obj.repr v);
-  let safe_r = identity r in
-  (* r._a is inlined by ocamlopt and equals None
-     because the field is supposed to be immutable. *)
-  assert (safe_r._a = v)
-
-let () = test ()
-
 (************************************)
