diff --git a/compiler/lib/ocaml_compiler.ml b/compiler/lib/ocaml_compiler.ml
index 29b6c8f058..93de84e68d 100644
--- a/compiler/lib/ocaml_compiler.ml
+++ b/compiler/lib/ocaml_compiler.ml
@@ -32,7 +32,7 @@ let rec constant_of_const : _ -> Code.constant =
   | Const_base (Const_int64 i) -> Int64 i
   | Const_base (Const_nativeint i) -> Int (Int32.of_nativeint_warning_on_overflow i)
   | Const_immstring s -> String s
-  | Const_float_array sl ->
+  | Const_float_array sl | Const_block_array sl ->
       let l = List.map ~f:(fun f -> Code.Float (float_of_string f)) sl in
       Tuple (Obj.double_array_tag, Array.of_list l, Unknown)
   | ((Const_pointer (i, _)) [@if BUCKLESCRIPT]) ->

